README
NAME
    App::tt - Time tracking application

VERSION
    0.03

DESCRIPTION
    App::tt is an application that can track how much time you spend on an
    project from command line.

    It is inspired by App::TimeTracker and share the same log file format,
    but it has (in my humble opinion) a simpler interface and easier to
    install.

SYNOPSIS
    The application is built up by specifying an an action and optional
    arguments.

      $ tt <action> [options]
      $ tt help <action>
      $ tt <action> -h

    Available actions: "analyze", "edit", "log", "register", "start",
    "status" (default) and "stop".

    Basic usage;

      # Start to track time
      $ cd $HOME/git/my-project
      $ tt start
      # Work, work, work, cd ..., do other stuff
      $ tt stop

    A bit more complex:

      # Start to work on an event and add a tag
      $ tt start -t ISSUE-999 -p some-project-at-work
      # Add another tag to the same event and add a --comment
      $ tt stop -t GITHUB-1005 "Today I was mostly in meetings"

ACTIONS
    Each action can tak "-h" for more details. Example:

      $ tt start -h

  analyze
    This action can analyze all the git repos in a directory and print what
    time you started/stopped working on a set of repos grouped per day. The
    output is suitable for "tt register".

      $ cd ..                  # one step up from the current git-project
      $ tt analyze             # default is last month
      $ tt analyze "last week" # analyze git reflog from last week
      $ tt analyze 2015-07-01  # analyze git reflog since a given date

  edit
    This command can be used to rewrite all the log entries. See source code
    before running this action. (Internals might change)

    DISCLAIMER! Backup your files before running this action!

      $ cat rewrite.pl | tt edit
      # verify output, and then run:
      $ cat rewrite.pl | tt edit --no-dry-run

  log
    This command will report how much time you have spent on various events.

      $ tt log         # this month
      $ tt log -2      # two months ago
      $ tt log year    # log for year
      $ tt log year -1 # last year
      $ tt log -p foo  # Filter by project name

  register
    This command is used to import data from other sources. "project-name"
    default to "-p" or current git project, "some description" default to
    "-d" and tags can be specified by -t foo -t bar

      $ tt register 2016-06-28T09:00:00 17:00:00 "project-name" "some description" "foo,bar"
      $ echo "2016-06-28T09:00:00\t2016-06-28T17:00:00\tproject-name\tsome description\tfoo,bar" | tt register

  start
    This command will start tracking a new event. It will also stop the
    current event if any event is in process. This action takes the "-p" and
    "-t" switches. "-p" (project) is not required if you start from a git
    repository.

      # Specify a tag and custom project name
      $ tt start -t ISSUE-999 some-project-name

      # Started working at 08:00 instead of now
      $ tt start 08:00

  status
    This is the default action and will return the current status: Are you
    working on something or not?

      $ tt status

  stop
    This command will stop tracking the current event.

      # Add more tags to the current event
      $ tt stop -t meetings -t not_fun

      # Stop working at 16:00 instead of now
      $ tt stop 16:00

COPYRIGHT AND LICENSE
    Copyright (C) 2014, Jan Henning Thorsen

    This program is free software, you can redistribute it and/or modify it
    under the terms of the Artistic License version 2.0.

AUTHOR
    Jan Henning Thorsen - "jhthorsen@cpan.org"



Changes
Revision history for perl distribution App-tt

0.19 Not Released
 - Changed to Getopt::App
 - Changed to group by day for "export" by default

0.18 2021-04-27T11:09:06+0900
 - Fix "tt edit file.trc"

0.17 2021-04-09T09:26:26+0900
 - Fix failing tests

0.16 2021-04-08T10:42:01+0900
 - Fix editing last file, even if "previous" is corrupted
 - Changed output text
 - Changed from JSON::XS to JSON::PP
 - Removed File::HomeDir as a dependency

0.15 2021-04-07T22:26:31+0900
 - Nicer formatting
 - Fix uninitialized duration
 - Can use default project with "register"

0.14 2020-12-05T09:58:22+0900
 - Fix Invalid time: 2020-13-01T00:00:00
 - Can read defaults from ~/.TimeTracker/config or $PWD/.env

0.13 2020-09-02T08:46:36+0900
 - Add support for "TT_COLUMNS=... tt export"
 - Can also edit description interactive
 - Mass edit can take --year and --month
 - Mass edit can be interactive

0.12 2020-04-02T15:36:31+0900
 - Add missing dependency List::Util 1.45+

0.11 2020-04-01T17:27:56+0900
 - Fix bug where editing would split tags on "-"
 - Fix not saving duplicate tags
 - Add current task to "tt log"
 - Add support for TT_HOURS_PER_MONTH
 - Changed "tt register" to require -p, -t, arguments
 - Changed "tt register" to not reading from STDIN
 - Changed time parsing to even more DWIM
 - Converted to "subcommand" from Applify

0.10 2019-02-04T09:35:24+0100
 - Fix "tt log -1" when running in January
 - Fix "tt log" when specifying interval

0.09 2018-09-28T11:17:05+0900
 - Add "export" action

0.08 2018-09-02T22:29:25+0200
 - Fix Can't locate object method "sysread" via package "IO::Handle" #3

0.07 2018-08-20T17:54:54+0200
 - Removed "analyze" action

0.06 2017-10-16T09:30:00+0200
 - Fix sorting "tt log" dates

0.05 2017-06-24T10:19:34+0200
 - Add support for "tt edit" or "tt edit <date>"

0.04 2016-09-26T09:02:00+0200
 - Fix setting default "stop" time HH:MM

0.03 2016-09-17T20:37:18+0200
 - Fix "register" help output
 - Allow register to only take hh:mm:ss as "stop"
 - Change "stop" action to use the same date as "start"

0.02 2016-07-19T11:02:15+0200
 - Fix double printing of messages
 - Fix uninitialized warnings
 - Make "register" command easier to copy/paste

0.01 2016-01-04T08:28:38+0100
 - Add action "analyze"
 - Add action "log"
 - Add action "register"
 - Add action "start" and "stop"
 - Add action "status"


lib/App/tt.pm
package App::tt;
use strict;
use warnings;

our $VERSION = '0.18';

1;

=encoding utf8

=head1 NAME

App::tt - Time tracking application

=head1 VERSION

0.18

=head1 DESCRIPTION

L<App::tt> is an application that can track how much time you spend on an
project from command line.

It is inspired by L<App::TimeTracker> and share the same log file format,
but it has (in my humble opinion) a simpler interface and easier to install.

=head1 SYNOPSIS

The application is built up by specifying an command and optional arguments.
Here is a list of example usages, but you can get more details by adding "-h"
after each command.

  # Start tracking time
  $ tt start
  $ tt start -p project-name
  $ tt start -p project-name -t tag1,tag2
  $ tt start -p project-name -t tag1,tag2 09:03

  # Current status
  $ tt
  $ tt status

  # Stop tracking time. Specifiying a time will go back to yesterday,
  # in case you forgot to stop it.
  $ tt stop
  $ tt stop 18:04

  # See the log
  $ tt log
  $ tt log -0year                  # Log for this year
  $ tt log -1year -t meetings      # Log for last year, for tag "meetings"
  $ tt log -p project-name -1month # Log for last month, for project "project-name"
  $ tt log -2                      # Log for two months back
  $ tt export -1                   # Export as CSV

  # Edit the last entry, or a specific file
  $ tt edit
  $ tt edit ~/.TimeTracker/2020/01/20200106-150000_nms.trc

  # Register forgotten time
  $ tt register 2020-01-01T09:00:00 17:00:00 -p "project-name"
  $ tt register 2020-01-01T09:00:00 17:00:00 -p "project-name" -t "tag1,tag2"
  $ tt register 2020-01-01T09:00:00 17:00:00 -p "project-name" -d "description" -t "tag1,tag2"

Basic usage;

  # Start to track time
  $ cd $HOME/git/my-project
  $ tt start

  # Work, work, work, cd ..., do other stuff
  $ tt stop

A more complex example:

  # Start to work on an event and add a tag
  $ tt start -t ISSUE-999 -p some-project-at-work

  # Add another tag to the same event and add a --comment and specify when
  # you stopped working
  $ tt stop -t GITHUB-1005 "Today I was mostly in meetings" 15:24

=head1 CONFIG

Default configuration can be read from either C<$PWD/.env> or
C<$HOME/.TimeTracker/config>. Here is an example config file with default
settings:

  # some comment
  editor=nano
  export_columns=date,project,hours,rounded,tags,description
  hours_per_month=0  # Used to calculate how much you have to work
  min_time=300       # Will drop the task on "tt stop" if started less than 300 seconds
  round_up_at=30     # Used by the export comnmand to round hours

Can also use the environment aliases:

  EDITOR=vim
  TT_COLUMNS=date,project
  TT_HOURS_PER_MONTH=150
  TT_MIN_TIME=600
  TT_ROUND_UP_AT=15

=head1 ACTIONS

Each command can take C<-h> for more details. Example:

  $ tt start -h

=head2 edit

This command can be used to rewrite a log entry.

  # Edit the last entry with your favorite $EDITOR
  $ tt edit

  # Edit a given file with your favorite $EDITOR
  $ tt edit ~/.TimeTracker/2017/12/20171220-092000_rg.trc

  # Rewrite all the log entries with a perl script
  # See source code before running this command. (Internals might change)
  $ cat rewrite.pl | tt edit

=head2 export

This will export a given set of records as CSV.

  $ tt export         # this month
  $ tt export -2      # two months ago
  $ tt export year    # log for year
  $ tt export -1y     # last year
  $ tt export -p foo  # Filter by project name

=head2 log

This command will report how much time you have spent on various
events.

  $ tt log         # this month
  $ tt log -2      # two months ago
  $ tt log year    # log for year
  $ tt log -1y     # last year
  $ tt log -p foo  # Filter by project name

If you set the "TT_HOURS_PER_MONTH" environment variable to the number of hours
you plan to work per month, then "tt log" will also print how many hours you
have to work in average to meet the target. Example:

  $ TT_HOURS_PER_MONTH=150 tt log -p my_job
  ...
  Remaining this month: 21d,  7:08h/d.

=head2 register

This command is used to import data from other sources. "project-name" default to
"-p" or current git project, "some description" default to "-d" and tags can be
specified by -t foo -t bar

  $ tt register 2020-01-01T09:00:00 17:00:00 -p project-name -d "some description" -t foo -t bar

=head2 start

This command will start tracking a new event. It will also stop the current
event if any event is in process. This command takes the "-p" and "-t"
switches. "-p" (project) is not required if you start from a git repository.

  # Specify a tag and custom project name
  $ tt start -t ISSUE-999 some-project-name

  # Started working at 08:00 instead of now
  $ tt start 08:00

=head2 status

This is the default command and will return the current status:
Are you working on something or not?

  $ tt status

=head2 stop

This command will stop tracking the current event.

  # Stop working at 16:00 instead of now
  $ tt stop 16:00

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014, Jan Henning Thorsen

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

=head1 AUTHOR

Jan Henning Thorsen - C<jhthorsen@cpan.org>

=cut
